packages/relayer-lib/src/tx_builder/eth_to_cosmos.rs

```diff
@@ -0,0 +1,429 @@
+
+                let computed_slot = compute_slot_at_timestamp(
+                    ethereum_client_state.genesis_time,
+                    ethereum_client_state.seconds_per_slot,
+                    std::time::SystemTime::now()
+                        .duration_since(std::time::UNIX_EPOCH)?
+                        .as_secs(),
+                )
+                .unwrap();
+                tracing::debug!(
+                    "Finality block number: {}, Update block number: {}, Update signature slot: {}, Target block number: {}, computed slot: {}",
+                    latest_finalized_block_number,
+                    latest_light_client_update_block_number,
+                    latest_ligth_client_signature_slot,
+                    target_block_number,
+                    computed_slot,
+                );
+                if latest_finalized_block_number > target_block_number
+                    && latest_light_client_update_block_number > target_block_number
+                    //&& target_period > trusted_period
+                    && computed_slot > latest_ligth_client_signature_slot
+                {
+                    return Ok(true);
+                }
+                Ok(false)
+            },
+        )
+        .await?;
+
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl<T, P> TxBuilderService<EthEureka, CosmosSdk> for TxBuilder<T, P>
+where
+    T: Transport + Clone,
+    P: Provider<T> + Clone,
+{
+    #[tracing::instrument(skip_all)]
+    async fn relay_events(
+        &self,
+        src_events: Vec<EurekaEvent>,
+        dest_events: Vec<EurekaEvent>,
+        target_channel_id: String,
+    ) -> Result<Vec<u8>> {
+        let target_block_number = self.eth_client.get_block_number().await?;
+        let channel = self.channel(target_channel_id.clone()).await?;
+
+        tracing::info!(
+            "Relaying events from Ethereum to Cosmos for channel {}",
+            target_channel_id
+        );
+        tracing::debug!("Target block number: {}", target_block_number);
+
+        let ethereum_client_state = self
+            .ethereum_client_state(channel.client_id.clone())
+            .await?;
+        let ethereum_consensus_state = self
+            .ethereum_consensus_state(channel.client_id.clone(), 0)
+            .await?;
+
+        self.wait_for_light_client_readiness(
+            target_block_number,
+            &ethereum_client_state,
+            &ethereum_consensus_state,
+        )
+        .await?;
+        let light_client_updates = self
+            .get_light_client_updates(&ethereum_client_state, &ethereum_consensus_state)
+            .await?;
+        tracing::debug!("light client updates: #{}", light_client_updates.len());
+
+        let mut headers = vec![];
+        let mut trusted_slot = ethereum_consensus_state.slot;
+        let mut prev_pub_agg_key = BlsPublicKey::default();
+        for update in &light_client_updates {
```

Comment:
I wonder if this for loop can be done with futures fold instead but this looks good for now